### PR TITLE
bake: initial implementation

### PR DESCRIPTION
**Pull request by tonistiigi**
Initial implementation for #13 (formely buildfile)

Testing:

```
> make
cd compose
docker buildx bake
docker buildx bake -f config.hcl release
docker buildx bake -f config.hcl --set db.args.FOO=bar release
# see config.hcl for full data format
# --print shows the parameters that would be used by individual builds without invoking it
```

The compose part is pretty basic and should work as expected. The full internal datastructure for defining targets probably needs some design work and doesn't look right to me.

Alternative designs:
```
group "default" {
	targets = ["db", "webapp"]
}


target "%-cross" {
	platforms = [
		"linux/amd64",
		"linux/arm64"
	]
}

---
docker buildx bake default-cross
```

```
target "default" {
	targets = ["db", "webapp"]
}

target "default-cross" {
	targets = ["db", "webapp-cross"]
}

target "webapp" {
	context = "./"
	dockerfile = "Dockerfile.webapp"
	args = {
		buildno = "123"
	}
}
```

```
group "default" {
	targets = ["db", "webapp"]
}

group "default-cross" {
        inherits = ["cross"]
	targets = ["db", "webapp"]
}
target "cross" {
	platforms = [
		"linux/amd64",
		"linux/arm64"
	]
}

---
docker buildx bake default-cross
```

Outputs not supported yet, thinking if this should be per target or global.

@tiborvass @AkihiroSuda @seemethere @dmcgowan

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Looks good, but why HCL?

**Comment by tonistiigi** [comment]
@AkihiroSuda I don't really care about the format that much. Could accept anything that we can parse. You can use JSON as well (and thinking if we should allow json5). Examples use HCL cause I personally found it more readable and better than toml on comparisons. I'm not a big fan of yaml but the main thing is that I'd want to avoid something that looks like compose but isn't compose. Atm. `bake` accepts vanilla compose with no changes and I think we should keep it this way. If you want you can use compose with the override logic as well. If you as a user like yaml and compose already you can just keep using that and never care about the other formats.

**Comment by tonistiigi** [comment]
updated buildkit to bring in moby/buildkit#929

File: bake/bake.go [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+	fnl := strings.ToLower(fn)
+	if strings.HasSuffix(fnl, ".yml") || strings.HasSuffix(fnl, ".yaml") {
+		return ParseCompose(dt)
+	}
+
+	if strings.HasSuffix(fnl, ".json") || strings.HasSuffix(fnl, ".hcl") {
+		return ParseHCL(dt)
+	}
+
+	cfg, err := ParseCompose(dt)
```
Comment (tiborvass) [inline bake/bake.go]: Are there compose files without `.yml` or `.yaml` suffixes? Why not error out?

Comment (tonistiigi) [reply bake/bake.go]: I have no idea. Does compose force you to use `.yml` suffix currently, I think not.

File: bake/bake.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+	bo := &build.Options{
+		Inputs: build.Inputs{
+			ContextPath:    t.Context,
+			DockerfilePath: t.Dockerfile,
+		},
+		Tags:      t.Tags,
+		BuildArgs: t.Args,
+		Labels:    t.Labels,
+		// CacheFrom: t.CacheFrom,
```
Comment (tiborvass) [inline bake/bake.go]: ?

Comment (tonistiigi) [reply bake/bake.go]: The key does not exist atm. but should later.

File: commands/bake.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+func runBake(dockerCli command.Cli, targets []string, in bakeOptions) error {
+	ctx := appcontext.Context()
+
+	if len(in.files) == 0 {
+		files, err := defaultFiles()
+		if err != nil {
+			return err
+		}
+		if len(files) == 0 {
+			return errors.Errorf("no compose.yml or dockerbuild.hcl found, speficy build file with -f/--file")
```
Comment (tiborvass) [inline commands/bake.go]: I think official is `docker-compose.yml`

Comment (tiborvass) [reply commands/bake.go]: typo: `specify`

Comment (tiborvass) [reply commands/bake.go]: `bakefile.hcl` ? :P

File: commands/bake.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	if err != nil {
+		return err
+	}
+
+	return buildTargets(ctx, dockerCli, bo, in.progress)
+}
+
+func defaultFiles() ([]string, error) {
+	fns := []string{
+		"compose.yml", // support app
```
Comment (tiborvass) [inline commands/bake.go]: we need `docker-compose.yml` too, and `.yaml` variants

File: commands/bake.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+		"dockerbuild.json",
+		"dockerbuild-override.json",
```
Comment (tiborvass) [inline commands/bake.go]: `dockerbuild.override.json`

**Comment by FernandoMiguel** [comment]
@tonistiigi what ever happened to `--set db.args.FOO=bar` ?

**Comment by tonistiigi** [comment]
@FernandoMiguel wdym?

**Comment by FernandoMiguel** [comment]
> @FernandoMiguel wdym?

@tonistiigi in the OP example
`docker buildx bake -f config.hcl --set db.args.FOO=bar release`

I'm trying to pass args to my buildx

**Pull request merged**